src/stacks/cloudfront_spa_pipeline_stack.ts

```diff
@@ -0,0 +1,80 @@
+import { CfnOutput, Construct } from "@aws-cdk/core";
+import BasePipelineStack, {
+  BasePipelineStackProps,
+} from "../base_stacks/base_pipeline_stack";
+import { PipelineProject, BuildSpec } from "@aws-cdk/aws-codebuild";
+import { Artifact } from "@aws-cdk/aws-codepipeline";
+import {
+  CodeBuildAction,
+  S3DeployAction,
+} from "@aws-cdk/aws-codepipeline-actions";
+import { HostedZone } from "@aws-cdk/aws-route53";
+import CloudfrontSpa from "../spa/cloudfront_spa";
+
+export interface CloudfrontSpaPipelineStackProps
+  extends BasePipelineStackProps {
+  subDomain?: string;
+  domainName: string;
+}
+
+export default class S3SpaPipelineStack extends BasePipelineStack {
+  constructor(
+    scope: Construct,
+    id: string,
+    props: CloudfrontSpaPipelineStackProps
+  ) {
+    super(scope, id, {
+      description: `Pipeline, Build & Deploy to Cloudfront for ${props.appName}-${props.appEnvironment} SPA Application`,
+      ...props,
+    });
+
+    const { subDomain = this.conventions.eqn("dash"), domainName } = props;
+
+    const hostedZone = HostedZone.fromLookup(scope, "Zone", {
+      domainName: domainName,
+    });
+
+    const spa = new CloudfrontSpa(this, "spa", {
```

Comment:
I think moving more logic into constructs or subclasses of constructs is the way to go - makes code more reusable.